### DISCUSSION TITLE
Make use of layout feature of TW to switch easily between horizontal and multicolumn layouts

### DISCUSSION THREAD

**Discussion by arunnbabu81** [crazko/krystal#50] [Ideas]
BTC is creating a [multicolumn layout for TW](https://burningtreec.github.io/tiddlywiki-multi-columns/). See the discussions [here](https://talk.tiddlywiki.org/t/a-multi-column-layout-storyview/2876). It would be nice if we could easily switch between the multicolumn and horizontal layout without much errors.

There is an interest in such an easy switching of layout in the TW community. See the discussions [here](https://talk.tiddlywiki.org/t/switching-theme-layout-storyview-with-one-click/3072).

**Comment by crazko** [comment]
I didn't find much info about `$:/layout` or `$:/tags/Layout`. Can you provide some source?

**Reply by arunnbabu81** [reply]
You may look at this discussion 

https://talk.tiddlywiki.org/t/is-there-a-button-to-switch-layout/3258

**Comment by crazko** [comment]
Adding http://szen.io/horizont/ by Jan Johannpeter for the reference.

**Comment by crazko** [comment]
Some discussion was in https://github.com/crazko/krystal/issues/54.

Some tutorial at https://alt-layout-tutorial.tiddlyhost.com/

**Comment by linonetwo** [comment]
@crazko I've update the layout tutorial https://tiddlywiki.com/#Alternative%20page%20layouts

Note that kyrstal has too many modification, or JS calls, so have conflicts with other plugins. Using a new layout will fix it.

Also see layouts like https://github.com/tiddly-gittly/health-buff-debuff-tracker/tree/master/src/health-buff-debuff-tracker/PageLayout

https://github.com/tiddly-gittly/TiddlyWiki-CPL/blob/master/tiddlers/CPLPlugin/layout.tid

**Comment by kookma** [comment]
I think the Krystal well lend itself to a TiddlyWiki layout! This way one can have both Krystal and Vanilla layout on one wiki at the same time!

**Comment by crazko** [comment]
Thank you for your tutorial update and other links! I'd like to look into that when the time allows it.

**Reply by linonetwo** [reply]
We are never hurry. My tutorial comes out when layout feature is already released for 4 years.

**Comment by linonetwo** [comment]
Oh, no, this is too hard. Because you use startup module to add logic, which is enabled no matter what story view type or layout user is using.

It's better to consider user may switch to other mode, instead of always using same storyview type or layout.

**Reply by linonetwo** [reply]
@crazko Is it possible to merge logic in `plugin.js` into `horizontal.js`? This can make sure logic unload after layout is quit.

**Reply by linonetwo** [reply]
Can't do that, `module-type: storyview` don't have lifecycle mechanism, can't register and unregister custom logics.

So I use a workarould, I add a `  const isInKrystalLayout = () => $tw.wiki.getTiddlerText("$:/layout") === "$:/plugins/rmnvsl/krystal/krystal-layout";`

and check it before every logic in `plugin.js`

**Comment by linonetwo** [comment]
![图片](https://github.com/user-attachments/assets/1fa32b96-bdd7-4494-aa65-7dc53125ee50)

![图片](https://github.com/user-attachments/assets/9c284189-8a9e-492d-b5cf-63abb2ebda4a)


Just done the refactor, take me 1 day. Have a try at [tiddly-gittly.github.io/krystal/](http://tiddly-gittly.github.io/krystal/)

Download the plugin JSON from https://github.com/tiddly-gittly/krystal/releases

The plugin is still under @crazko 's name `$:/plugins/rmnvsl/krystal`. But @crazko  if you think the change is too large, I could release it under my name as a new plugin, while still credit that it is your work.

**Reply by pmario** [reply]
The first link does not work

**Reply by linonetwo** [reply]
@pmario Updated, I move it to https://github.com/tiddly-gittly/krystal , because now instead of PR, I fork it and release it as a new plugin, because change set is too large, discussed in https://github.com/crazko/krystal/pull/60#issuecomment-2726790680
